# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/new-regexp/regexp-shim.h
# Commit: 534658e99952
# Full Hash: 534658e9995271198d6a71278787d8aa6e02e09f
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2020-05-13 21:46:27
# Regressor Bug: 1634135
# File Overlap Count: 1
# Description:
#   Bug 1634135: Update shim code r=mgaudet
#   
#   There are a few changes here:
#   
#   1. The code that used to be in wrapBody was moved to RegExpCompiler::PreprocessRegExp.
# ==============================================================================

diff -r 09b8bab47f6b -r 534658e99952 js/src/new-regexp/regexp-shim.h
--- a/js/src/new-regexp/regexp-shim.h	Wed May 13 16:22:44 2020 +0000
+++ b/js/src/new-regexp/regexp-shim.h	Wed May 13 16:22:46 2020 +0000
@@ -161,6 +161,19 @@
   return (x >= 0) ? ((x < 255) ? uint8_t(x) : 255) : 0;
 }
 
+// Origin: https://github.com/v8/v8/blob/fc088cdaccadede84886eee881e67af9db53669a/src/base/bounds.h#L14-L28
+// Checks if value is in range [lower_limit, higher_limit] using a single
+// branch.
+template <typename T, typename U>
+inline constexpr bool IsInRange(T value, U lower_limit, U higher_limit) {
+  using unsigned_T = typename std::make_unsigned<T>::type;
+  // Use static_cast to support enum classes.
+  return static_cast<unsigned_T>(static_cast<unsigned_T>(value) -
+                                 static_cast<unsigned_T>(lower_limit)) <=
+         static_cast<unsigned_T>(static_cast<unsigned_T>(higher_limit) -
+                                 static_cast<unsigned_T>(lower_limit));
+}
+
 #define LAZY_INSTANCE_INITIALIZER { mozilla::Nothing() }
 
 template <typename T>
@@ -506,6 +519,8 @@
     return JS::Value::fromRawBits(asBits_);
   }
 
+  inline static Object cast(Object object) { return object; }
+
  protected:
   void setValue(const JS::Value& val) {
     asBits_ = val.asRawBits();
@@ -636,11 +651,6 @@
             typename = std::enable_if_t<std::is_convertible_v<S*, T*>>>
   inline Handle(Handle<S> handle) : location_(handle.location_) {}
 
-  template <typename S>
-  inline static const Handle<T> cast(Handle<S> that) {
-    return Handle<T>(that.location_);
-  }
-
   inline bool is_null() const { return location_ == nullptr; }
 
   inline T operator*() const {
@@ -906,6 +916,13 @@
     return regexp;
   }
 
+  // Each capture (including the match itself) needs two registers.
+  static int RegistersForCaptureCount(int count) { return (count + 1) * 2; }
+
+  inline int MaxRegisterCount() const {
+    return inner()->getMaxRegisters();
+  }
+
   // ******************************
   // Static constants
   // ******************************